corehq/apps/userreports/reports/builder/forms.py

```diff
@@ -567,6 +572,18 @@ def _get_owner_name_pseudo_property():
             data_types=["string"],
         )
 
+    @classmethod
+    def _get_owner_location_pseudo_property(cls):
+        # owner_location is a special pseudo-case property for which
+        # the report builder reference the owner_id, but treat it as a location
+        return DataSourceProperty(
+            type=PROPERTY_TYPE_CASE_PROP,
+            id=COMPUTED_OWNER_LOCATION_PROPERTY_ID,
+            text=_('Case Owner (Location)'),
```

Comment:
Does this only work if the owner is a Location? Can it use the location of the owner if the owner is a CommCareUser?

Reply:
I think that's basically what the current `Case Owner` property does but that one isn't location-safe.

Note that I don't think this is the right productized way to do this, but wanted something available as a better option than jumping into a manual UCR.

In the productized version I'd think ideally there would be a way to use the same property for both and either:

1. Somehow specify in the filter that you want to model it as a location with location safety (we don't really have the UI setup in such a way as to facilitate that easily right now)
2. Automatically make it location-safe by somehow understanding the relationships between users, groups (maybe hard) and locations.

I think @esoergel mentioned a spec for this, so am also curious what they had in mind. I definitely just went with the lowest-effort approach to get something working.

Reply:
I'm actually not aware of a spec for this.

> Automatically make it location-safe by somehow understanding the relationships between users, groups (maybe hard) and locations.

The standard product reports are a good paradigm to follow here.  Generally how they work is that location-restricted users don't have access to groups at all (they don't play well with locations), but they do have access to users - only those who are currently assigned to a location they have access to.  [Here's where the set of filter options is defined](https://github.com/dimagi/commcare-hq/blob/22d128b06b17ef3ddd9fa5c4c347ec244e8d6cec/corehq/apps/reports/filters/controllers.py#L257-L261).

Here's the in-memory join (:vomiting_face:) used to filter [the set of accessible users](https://github.com/dimagi/commcare-hq/blob/22d128b06b17ef3ddd9fa5c4c347ec244e8d6cec/corehq/apps/reports/filters/controllers.py#L117-L120) for the filter.

Note also that this differs depending on whether you're filtering forms or cases.  Forms are only associated with users, so you need to show appropriate filter options, then get all users matching those filters (whether directly selected users or users at or below selected locations).  [This util does that](https://github.com/dimagi/commcare-hq/blob/22d128b06b17ef3ddd9fa5c4c347ec244e8d6cec/corehq/apps/reports/filters/users.py#L341-L341).  Cases can be owned by users, defined in the same way, or directly by locations, so getting the set of case owners is a bit [more complicated](https://github.com/dimagi/commcare-hq/blob/22d128b06b17ef3ddd9fa5c4c347ec244e8d6cec/corehq/apps/reports/standard/cases/utils.py#L62-L62).

Last thing I'll note on this is my mental model around these things.  If a report displays forms or cases, HQ has a paradigm for that, and consistency is the obvious choice.  In UCRs, however, rows represent project-specific concepts, which may not map as cleanly.  For those, we leave it to the user to define how each row is associated with a location.  Report Builder reports are much more restricted, so maybe it does make sense to just treat them as forms or cases, which would let us reuse concepts (and utils?) from the standard reports.

Reply:
Wow, thanks this is awesome feedback (almost wonder if it should be in the docs or something if it's not already). That said, it does sound rather complex and I think my time is going to be better spent elsewhere at the moment so will go ahead with this option for now.

Also I was confused, apparently (according to Erin on field/dev) it's @proteusvacuum that had the spec...

Reply:
Thanks!  I'm not sure that's laid out in docs anywhere - where would you look for this sort of thing?

Reply:
Yeah, not 100% sure. Somewhere in [this general vicinity](https://commcare-hq.readthedocs.io/locations.html#location-permissions) would be my first guess. Not sure it's worth it, it just seems a shame to have such useful information buried in PR comment that no one else will ever find.

Reply:
sorry for the delay - here are those docs: https://github.com/dimagi/commcare-hq/pull/27398

Reply:
This is probably not very helpful, but here is the spec (last edited in 2018...) that I think Erin was referring to: https://docs.google.com/document/d/18ip_jgsQZTzXhQXu-wpErloaBUp22MB5vev_Dqo52lM/edit#

Reply:
thanks. most of the implementation was similar, though the spec has several usability improvements. I added a link and list here: https://confluence.dimagi.com/display/ccinternal/Enable+creation+of+report+builder+reports+that+are+location+safe